Search updates

rtibbles opened the pull request:
## Summary
* Bootstraps all available channel data into the learn page
* Uses this throughout the learn plugin to avoid API calls
* Creates a composable helper for search functionality
* Creates a SearchChips component to display and translate currently selected search terms
* Adds a route for searching within topics and uses route state for most behaviour on topics page
* Gets rid of TOPICS_CHANNEL route in favour of just using TOPICS_TOPIC
* Fixes some issues with breadcrumbs
* Makes search in topics constrained to descendants of the currently navigated to topic
* Deletes unused code that referenced the now non-existent TOPICS_ROOT and SEARCH routes
* Upgrades kolibri-constants and adds additional unhandled translations to core translator

## References
Fixes #8565
Fixes #8555
Fixes #8556
Fixes #8557
Also resolves some TODOs from #8563:
* Remove breadcrumbs for search
* Remove "channels" breadcrumb (although changes to Library - maybe should remove completely?)

## Reviewer guidance
![search](https://user-images.githubusercontent.com/1680573/139608913-8f254f8b-535c-47ca-a57e-6d4bd2455480.gif)


----

## Testing checklist

- [x] Contributor has fully tested the PR manually
- [x] If there are any front-end changes, before/after screenshots are included
- [x] Critical user journeys are covered by Gherkin stories
- [ ] Critical and brittle code paths are covered by unit tests


## PR process

- [x] PR has the correct target branch and milestone
- [x] PR has 'needs review' or 'work-in-progress' label
- [x] If PR is ready for review, a reviewer has been added. (Don't use 'Assignees')
- [ ] If this is an important user-facing change, PR or related issue has a 'changelog' label
- [ ] If this includes an internal dependency change, a link to the diff is provided

## Reviewer checklist

- Automated test coverage is satisfactory
- PR is fully functional
- PR has been tested for [accessibility regressions](http://kolibri-dev.readthedocs.io/en/develop/manual_testing.html#accessibility-a11y-testing)
- External dependency files were updated if necessary (`yarn` and `pip`)
- Documentation is updated
- Contributor is in AUTHORS.md

rtibbles committed:
Use DjangoJSONEncoder for bootstrapping plugin data to allow datetimes.

rtibbles committed:
Bootstrap full available channel data into the learn page.

rtibbles committed:
Update useChannels composable to use bootstrapped channel data.
Update all channel references to use useChannels.

rtibbles committed:
Consolidate search logic in useSearch composable.
Consolidate search chips in SearchChips component.

rtibbles committed:
Don't show channels selector on within channel search.
Cleanup nonexistent prop.

rtibbles committed:
Fix channel breadcrumbs display.

rtibbles committed:
Generalize channel specificity for search to topic specificity.

rtibbles committed:
Simplify routes. Add topics search route.

rtibbles committed:
Fix search chips display. Add translation for no categories.

rtibbles committed:
Fix search clearing.

rtibbles committed:
Consolidate library page handler into a single function.

rtibbles committed:
Fix breadcrumbs for route changes.

rtibbles committed:
Clean up unused constants, components and functionality.

rtibbles committed:
Upgrade kolibri-constants.
Clean up lock file.
Add additional nonconforming keys.

rtibbles committed:
Add code comments to convoluted categories nesting code.

MisRob commented on kolibri/plugins/learn/assets/src/composables/useSearch.js:
```diff
@@ -0,0 +1,173 @@
+
+  watch(searchTerms, search);
+
+  return {
```
Since it's quite critical logic used from more places, would be good to have this composable's computed and methods that are publicly exposed here covered by tests, at least those that are more complex.  I realize that even without that this PR is a step towards reducing some tech debt, so it's fine to open a follow-up issue.

rtibbles replied:
Yes, some tests would be good, I think not unreasonable to do them within this PR.

rtibbles replied:
Have now added tests for every branch in here, I think.

MisRob commented on kolibri/plugins/learn/assets/src/views/TopicsPage.vue:
```diff
@@ -519,31 +508,18 @@
     created() {
+      this.translator = crossComponentTranslator(LibraryPage);
       window.addEventListener('scroll', this.throttledHandleScroll);
+      this.setSearchWithinDescendant(this.topic);
```
Since compostables can be used from anywhere, I'd suggest also having `resetSearchWithinDescendant` function available in the `useSearch` composable that would reset [`descendant`'s value here](https://github.com/learningequality/kolibri/pull/8570/files#diff-da88993a9f71c968b198aaca81c2fe54512de05371c08f6620e1a5dea267502dR28) to `undefined` and calling the reset somewhere in the `TopicsPage` after finishing all search, maybe in the `beforeDestroy` hook? Alternatively we could just call `setSearchWithinDescendant(undefined)`.

rtibbles replied:
All the data is local to the invocation of the composable though, so this won't get reused elsewhere.

MisRob replied:
Ah, that's right, it's in the function, not outside of it. I didn't realize that when writing the comment, you can ignore it then.

MisRob commented on kolibri/plugins/learn/assets/src/views/TopicsPage.vue:
```diff
@@ -519,31 +508,18 @@
-        );
-
-        if (contentNodeIds.length > 0) {
-          ContentNodeProgressResource.fetchCollection({ getParams: { ids: contentNodeIds } }).then(
```
What was the reason for removing fetching content nodes' progress data? It seems that `HybridLearningContentCard` on `TopicsPage` is prepared to display progress and I haven't noticed an alternative API call in this PR diff.

rtibbles replied:
I think this may have been a half finished thought - will follow up.

rtibbles replied:
Ah yes, this was intentional. The content nodes for which progress is being fetched here are not rendered anywhere on this page, this seemed to have been copy pasted from the `created` implementation in the `LibraryPage.vue` so was a superfluous request.

It does mean we still need to load progress for the items on the topics page, but this change is not a regression.

MisRob reviewed (dismissed):
Thank you, @rtibbles. I haven't managed to understand and review all parts of this PR yet. I added some comments to parts that I went through in more detail and I also see the following issues:

**1. I can't remove search by clicking on the delete button in a search chip**

![bug-chip](https://user-images.githubusercontent.com/13509191/139669179-fd85dd1b-c2d8-49dd-8796-9a57711faa86.gif)

**2. Navigation from a channel card to a channel page on the home page stopped working**
![bug-navigation](https://user-images.githubusercontent.com/13509191/139667050-8bf46223-4b06-4300-90b5-01cd32421984.gif)

**3. Home page and classes pages content area is wider, is that change intentional?**

| Before | After |
| -------- | ------- |
| ![classes-page-before](https://user-images.githubusercontent.com/13509191/139668159-7edb3fe3-3094-44b9-b774-fd7967dc1149.png) | ![classes-page-after](https://user-images.githubusercontent.com/13509191/139668183-10795493-6cb0-4c1a-a207-a0b5d2859d8d.png) |
| ![class-page-before](https://user-images.githubusercontent.com/13509191/139668204-4224accf-f151-4aff-ab03-2e8048950c0d.png) | ![class-page-after](https://user-images.githubusercontent.com/13509191/139668235-a60a4499-88c3-4df4-93c2-33406baf0a47.png) |
| ![home-page-before](https://user-images.githubusercontent.com/13509191/139668263-e435848d-b74b-4623-be1c-d704a59973d2.png) | ![home-page-after](https://user-images.githubusercontent.com/13509191/139668277-a348c33c-6d0c-459c-9c29-b50d3071c30d.png) |

rtibbles commented:
Thanks @MisRob:

> 1. I can't remove search by clicking on the delete button in a search chip

Looks like this was specific to key word searches only. Was unable to replicate with a non-keyword search item.

> 2. Navigation from a channel card to a channel page on the home page stopped working

This is probably related to the route simplification, will fix.

> 3. Home page and classes pages content area is wider, is that change intentional?

This was not intentional - it's the result of slightly overzealous LearnIndex cleanup. Will revert.

marcellamaki reviewed (changes requested):
I am going to do a closer code read-through (I've done a quick pass to understand the general changes). Overall the changes look good, although I appreciate that @MisRob is able to weigh in on the composables because I'm not sure I would things there. 

The one bug I've noticed in manual QA is that the options for the filter options persist the previous state after clearing the search:
![filters-persist](https://user-images.githubusercontent.com/17235236/139705160-5443df6d-83db-4dff-9626-f0cd428bb2b3.gif)

Will update with additional questions/comments this afternoon after a closer code review.

rtibbles commented:
Ah yes - good catch, can clear labels too on search change.

rtibbles committed:
Handle search clearing properly in all cases.
Remove width setting only used by non-existent route.
Remove TOPICS_CHANNEL and TOPICS_ROOT references.
Fix JS tests.

marcellamaki commented on kolibri/plugins/learn/assets/src/modules/topicsTree/handlers.js:
```diff
@@ -1,37 +1,27 @@
-  ConditionalPromise.all(promises).only(
+  ContentNodeResource.fetchModel({ id }).only(
     samePageCheckGenerator(store),
-    ([content, nextContent]) => {
```
I can't totally tell if this is a refactor, or a removal of "next content". Does any of @nucleogenesis 's side panel work depend on this for the topics/lesson overview side panel?

rtibbles replied:
My recollection was that this was being independently fetched in the side panel, and so it was now superfluous here.

I think that is still correct, as I see the recommendations being loaded here: https://github.com/learningequality/kolibri/blob/release-v0.15.x/kolibri/plugins/learn/assets/src/views/BrowseResourceMetadata.vue#L298

marcellamaki commented on kolibri/plugins/learn/assets/src/views/SearchChips.vue:
```diff
@@ -0,0 +1,132 @@
+    },
+    created() {
+      const LibraryPageComponent = require('./LibraryPage').default;
+      this.translator = crossComponentTranslator(LibraryPageComponent);
```
This is new!

rtibbles replied:
Yes, there were strings required to display the search chips that did not exist in the TopicsPage component, only in the LibraryPage component - but because we import the SearchChips into the LibraryPage, I didn't want to setup a circular import, so did the import here inside the closure.

rtibbles replied:
And also, I hate myself for using crossComponentTranslator.

marcellamaki commented on kolibri/plugins/learn/assets/src/routes/index.js:
```diff
@@ -99,23 +89,23 @@ export default [
   },
   {
-    name: PageNames.TOPICS_CHANNEL,
-    path: '/topics/:channel_id',
```
This feels like a notable change. Not blocking, and not a technical comment, but I'm wondering if this might require some toolkit or other updates for public channels to places like on the demo server, etc? I'm not sure how many places we have links to specific channel resources but thought that might be worth flagging

rtibbles replied:
I think it would probably be a good idea to retain the route, and instead do a redirect to the new page, in the interests of the above.

rtibbles replied:
Have added the redirect.

rtibbles committed:
Rename tree viewset parameter to avoid collision with contentnode filter.

marcellamaki commented on kolibri/plugins/learn/assets/src/views/Breadcrumbs.vue:
```diff
@@ -48,67 +58,25 @@
-            text: this.coreString('channelsLabel'),
-            link: { name: PageNames.TOPICS_ROOT },
+            text: this.learnString('libraryLabel'),
+            link: { name: PageNames.LIBRARY },
```
Breadcrumbs in the topics page don't need to start until the user is down at least one folder level. No reference to Library page is needed, just starting with the return to the highest topic (channel) folder. Which I only know because Peter already provided me this feedback 😄 
<img width="328" alt="Screen Shot 2021-11-01 at 3 38 55 PM" src="https://user-images.githubusercontent.com/17235236/139731186-1b32d949-5e19-4d38-93ff-e39dbe9248c6.png">

rtibbles replied:
Have updated to remove the Library from the breadcrumbs.

rtibbles committed:
Add redirect to handle historic channels page.
Remove breadcrumbs reference to Library.

rtibbles commented:
Resolved review feedback, except for adding tests for the search composable.

rtibbles committed:
Add full test coverage to useSearch module. Fix bugs and cleanup.

rtibbles commented:
Tests for search composable have now been added.

MisRob commented on kolibri/plugins/learn/assets/src/composables/__tests__/useSearch.spec.js:
```diff
@@ -0,0 +1,460 @@
+const name = 'not important';
+
+function prep(query = {}) {
+  const store = coreStoreFactory({
```
No action is needed. Just wanted to note that importing the whole store machinery can often bring quite a lot of complexity behind scenes and am not sure if it's worth it only to mock `state.route`, especially in a composable that is otherwise quite independent of the rest of the store. Was there any particular reason for bringing the whole core module into tests when compared to using an empty mock store or router and defining only values and methods that are needed?

rtibbles replied:
I did this because it's more representative of what is actually happening in production, and I didn't want any weird surprises to happen there compared to the tests.

Ideally, I would have just used the [VueRouter composable functions](https://next.router.vuejs.org/guide/advanced/composition-api.html) in here to access the route and the router, and not had the complexity inside or outside the tests, but that requires upgrading to VueRouter 4.x, and that seemed like a step too far for now!

MisRob replied:
> I did this because it's more representative of what is actually happening in production, and I didn't want any weird surprises to happen there compared to the tests

Thank you for the explanation, makes sense. I usually find balancing these things quite tricky. It sounds that this approach has the advantage of being more representative as you say.

MisRob replied:
No need to upgrade Vue Router for this :smile:

marcellamaki commented:
Seems like feedback has been addressed, although I know @jtamiace sent some feedback about the homepage -- not sure if that will be included in this PR or not. Otherwise, the only thing I see is this "extra chip" popping up. Seems like one empty one is appended to the end of the list of filters. 

<img width="445" alt="Screen Shot 2021-11-02 at 10 40 38 AM" src="https://user-images.githubusercontent.com/17235236/139872687-c6d8c97a-6eeb-493e-bbb1-5fbdd61b6f20.png">

rtibbles commented:
> not sure if that will be included in this PR or not

I think best to do as a follow up.

The extra chip is probably coming from an empty `keywords` parameter, will update.

rtibbles committed:
Don't display chips for null or empty keywords value.

rtibbles commented:
> The extra chip is probably coming from an empty keywords parameter, will update.

Replicated and updated.

nucleogenesis reviewed (commented):
Code look good. I really love seeing the composition API stuff in action and seeing it mixed with Vuex stuff - love it. I didn't see anything that jumped out at me. 

I found a bug in the keywords processing in Python which is that it fails with the following queries, separated by line:

```
10
100
hello 10
hello 100
10 hello
10 x 10
10 10 10
100 10 100
or
and or
```

These are all unlikely queries, of course, but they may signal that there are more that I'm not finding. I wouldn't block merging on this but it's worth following up on, perhaps after bug bash to see if people run into any more normal queries that return 500.

The server gives a message from Django that it [needs to be valid JSON (traceback)](https://gist.github.com/nucleogenesis/d0a8ac74658ee531df8312170c4ae8a3).

In any case - the fail state there when it gets the 500 is to show a spinner forever so that could use some better error handling for the UX - maybe just showing 0 results along with some existing "we encountered an error" strings. The user can return to normal by clearing the filters that appear once you click something in the side bar.

nucleogenesis reviewed (commented):
Oh just noticed this as I was wrapping up but the X shows in the search bar when you haven't searched any kind of text query.

So - I click "Read" and the X shows in both searches (top bar and left nav). 

When I click it, nothing happens. This is actually a bug altogether - even when I type text in and click the X to clear the search, it doesn't have any effect on the chips or query results.

If this isn't in scope I can make sure to create a follow up issue.

rtibbles committed:
Update searchInputValue to always be empty string for no value.

rtibbles commented:
Fixed the 500 in #8579

Have fixed the search box `X` issue:

![Screenshot from 2021-11-02 14-39-45](https://user-images.githubusercontent.com/1680573/139955617-f6b3b446-28cd-4c39-95c1-4d0c09eae358.png)

![Screenshot from 2021-11-02 14-39-50](https://user-images.githubusercontent.com/1680573/139955625-cb6862cf-1051-4a32-a6c9-46100e193858.png)
